src/celeritas/io/RootStepWriter.hh

```diff
@@ -0,0 +1,120 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/io/RootStepWriter.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <array>
+
+#include "celeritas_config.h"
+#include "corecel/Assert.hh"
+#include "celeritas/io/RootFileManager.hh"
+#include "celeritas/phys/ParticleParams.hh"
+#include "celeritas/user/StepInterface.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Write "MC truth" data to ROOT at every step.
+ *
+ * `TTree::Fill()` is called for each step and thread id, making each ROOT
+ * entry a step. Since the ROOT data is stored in branches with primitive types
+ * instead of a full struct, no dictionaries are needed for reading the output
+ * file.
+ */
+class RootStepWriter final : public StepInterface
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using SPRootFileManager = std::shared_ptr<RootFileManager>;
+    using SPParticleParams  = std::shared_ptr<const ParticleParams>;
+    //!@}
+
+    // Construct with RootFileManager and ParticleParams
+    explicit RootStepWriter(SPRootFileManager root_manager,
+                            SPParticleParams  particle_params,
+                            StepSelection     selection,
+                            Filters           filters);
+
+    // Set number of entries stored in memory before being flushed to disk
+    void set_auto_flush(long num_entries);
+
+    // Process step data on the host and fill step tree
+    void execute(StateHostRef const& steps) final;
+
+    // Device execution cannot be implemented
+    void execute(StateDeviceRef const&) final
+    {
+        CELER_NOT_IMPLEMENTED("RootStepWriter is host-only.");
+    }
+
+    // Selection of data to be stored
+    StepSelection selection() const final { return selection_; }
+
+    // Selection of filters for the stored data
+    Filters filters() const final { return filters_; }
+
+  private:
+    // Create steps tree based on selection_ booleans
+    void make_tree();
+
+  private:
+    //// INPUT ////
+    SPRootFileManager root_manager_;
+    SPParticleParams  particles_;
+    StepSelection     selection_;
+    Filters           filters_;
+
+    //// DATA ////
+    detail::TRootUniquePtr<TTree> tstep_tree_;
```

Comment:
I'd reorganize your private stuff data so it's all in a single block:
```
  private:
    //// TYPES ////
(classes here)

    //// DATA ////
(input, data, tstep)

    //// FUNCTIONS ////
(your private implementation helpers)
```

Definitely keep member data together because of how C++ does initialization (you might need to set and order them in the constructor list).